org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/WorkspaceFolderChangeHandler.java

```diff
@@ -44,6 +49,24 @@ public void update(DidChangeWorkspaceFoldersParams params) {
 				removedRootPaths.add(rootPath);
 			}
 		}
+		updateRootPaths(addedRootPaths, removedRootPaths);
 		projectManager.updateWorkspaceFolders(addedRootPaths, removedRootPaths);
 	}
+
+	private void updateRootPaths(Collection<IPath> addedRootPaths, Collection<IPath> removedRootPaths) {
+		Set<IPath> rootPathSet = new HashSet<>(preferenceManager.getPreferences().getRootPaths());
+
+		Set<IPath> removedSet = new HashSet<>(removedRootPaths);
+		for (IPath removed : removedSet) {
+			if (rootPathSet.contains(removed)) {
+				rootPathSet.remove(removed);
+			}
+		}
+
+		Set<IPath> addedSet = new HashSet<>(addedRootPaths);
```

Comment:
why not use `rootPathSet.addAll()` api directly?

Reply:
yes, that's much better